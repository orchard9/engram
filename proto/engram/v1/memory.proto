syntax = "proto3";

package engram.v1;

import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";

// ============================================================================
// Confidence Representation - First-class uncertainty quantification
// ============================================================================

// Confidence represents probabilistic certainty with both numeric and semantic categories.
// NEVER make confidence optional - absence of uncertainty info creates cognitive traps.
// Based on Gigerenzer & Hoffrage research on qualitative vs numeric uncertainty.
message Confidence {
  // Numeric confidence value in range [0.0, 1.0]
  float value = 1;
  
  // Semantic category for natural reasoning (auto-derived from value)
  ConfidenceCategory category = 2;
  
  // Optional explanation of confidence source
  string reasoning = 3;
}

// Semantic confidence categories that match human intuition
enum ConfidenceCategory {
  CONFIDENCE_CATEGORY_UNSPECIFIED = 0;
  CONFIDENCE_CATEGORY_NONE = 1;        // 0.0 - Complete uncertainty
  CONFIDENCE_CATEGORY_LOW = 2;         // 0.1 - "Probably not"
  CONFIDENCE_CATEGORY_MEDIUM = 3;      // 0.5 - "Maybe" / "Unsure"  
  CONFIDENCE_CATEGORY_HIGH = 4;        // 0.9 - "Quite sure"
  CONFIDENCE_CATEGORY_CERTAIN = 5;     // 1.0 - Complete certainty
}

// ============================================================================
// Core Memory Types - Psychological memory systems vocabulary
// ============================================================================

// Memory represents a single memory trace in the probabilistic graph.
// Follows forgetting curve research with activation and decay properties.
message Memory {
  // Core fields (1-9) - fundamental memory properties
  string id = 1;                             // Unique identifier
  repeated float embedding = 2;              // 768-dimensional semantic vector
  float activation = 3;                      // Current activation level [0.0, 1.0]
  Confidence confidence = 4;                 // Cognitive confidence (REQUIRED)
  google.protobuf.Timestamp last_access = 5; // When last recalled
  google.protobuf.Timestamp created_at = 6;  // When encoded
  float decay_rate = 7;                      // Forgetting curve parameter
  
  // Extended fields (10-19) - additional properties
  string content = 10;                       // Optional human-readable content
  map<string, string> metadata = 11;         // Extensible metadata
  
  // Advanced fields (20+) - specialized properties
  repeated string tags = 20;                 // Semantic tags
  MemoryType memory_type = 21;               // Type categorization
}

// Memory type categorization based on psychological research
enum MemoryType {
  MEMORY_TYPE_UNSPECIFIED = 0;
  MEMORY_TYPE_EPISODIC = 1;      // Specific events with context
  MEMORY_TYPE_SEMANTIC = 2;      // Facts and general knowledge
  MEMORY_TYPE_PROCEDURAL = 3;    // Skills and how-to knowledge
  MEMORY_TYPE_WORKING = 4;        // Temporary active memory
}

// Episode represents an episodic memory with rich contextual information.
// Based on Tulving's episodic memory research - "mental time travel".
message Episode {
  // Core episodic fields (1-9)
  string id = 1;                              // Unique episode identifier
  google.protobuf.Timestamp when = 2;         // When it happened
  string what = 3;                            // What happened (semantic content)
  repeated float embedding = 4;               // 768-dimensional representation
  Confidence encoding_confidence = 5;         // Encoding quality (REQUIRED)
  
  // Context fields (10-19)
  string where_location = 10;                 // Where it occurred
  repeated string who = 11;                   // Who was involved
  string why = 12;                            // Why it happened (causal)
  string how = 13;                            // How it happened (procedural)
  
  // Memory dynamics (20-29)
  float decay_rate = 20;                      // Episode-specific forgetting
  float emotional_valence = 21;               // Emotional charge [-1, 1]
  float importance = 22;                      // Subjective importance [0, 1]
  
  // Consolidation fields (30+)
  ConsolidationState consolidation_state = 30;
  google.protobuf.Timestamp last_replay = 31; // Last consolidation time
}

// Consolidation states following sleep/memory research
enum ConsolidationState {
  CONSOLIDATION_STATE_UNSPECIFIED = 0;
  CONSOLIDATION_STATE_RECENT = 1;      // Recently encoded, not consolidated
  CONSOLIDATION_STATE_CONSOLIDATING = 2; // Currently being consolidated
  CONSOLIDATION_STATE_CONSOLIDATED = 3;  // Stable long-term memory
  CONSOLIDATION_STATE_RECONSOLIDATING = 4; // Being updated/modified
}

// ============================================================================
// Retrieval Cues - Different ways to trigger memory recall
// ============================================================================

// Cue represents a memory retrieval trigger with multiple modalities.
// Supports pattern completion and spreading activation.
message Cue {
  // Cue identification
  string id = 1;
  
  // Cue modality (oneof ensures single cue type)
  oneof cue_type {
    EmbeddingCue embedding = 2;
    ContextCue context = 3;
    TemporalCue temporal = 4;
    SemanticCue semantic = 5;
    PatternCue pattern = 6;
  }
  
  // Retrieval parameters
  Confidence threshold = 10;           // Minimum confidence for recall (REQUIRED)
  int32 max_results = 11;              // Maximum memories to retrieve
  bool spread_activation = 12;         // Enable spreading activation
  float activation_decay = 13;         // Activation decay during spreading
}

// Embedding-based similarity search
message EmbeddingCue {
  repeated float vector = 1;           // 768-dimensional query vector
  float similarity_threshold = 2;      // Cosine similarity threshold
}

// Context-based retrieval (episodic)
message ContextCue {
  google.protobuf.Timestamp time_start = 1;
  google.protobuf.Timestamp time_end = 2;
  string location = 3;
  repeated string participants = 4;
}

// Temporal pattern search
message TemporalCue {
  TemporalPattern pattern = 1;
  google.protobuf.Timestamp reference_time = 2;
}

// Temporal patterns for memory search
message TemporalPattern {
  oneof pattern {
    string cron_expression = 1;        // Cron-like pattern
    RepeatingInterval interval = 2;    // Simple repetition
    RelativeTime relative = 3;         // Relative to reference
  }
}

message RepeatingInterval {
  int32 value = 1;
  TimeUnit unit = 2;
}

message RelativeTime {
  int32 offset = 1;
  TimeUnit unit = 2;
  RelativeDirection direction = 3;
}

enum TimeUnit {
  TIME_UNIT_UNSPECIFIED = 0;
  TIME_UNIT_SECONDS = 1;
  TIME_UNIT_MINUTES = 2;
  TIME_UNIT_HOURS = 3;
  TIME_UNIT_DAYS = 4;
  TIME_UNIT_WEEKS = 5;
  TIME_UNIT_MONTHS = 6;
  TIME_UNIT_YEARS = 7;
}

enum RelativeDirection {
  RELATIVE_DIRECTION_UNSPECIFIED = 0;
  RELATIVE_DIRECTION_BEFORE = 1;
  RELATIVE_DIRECTION_AFTER = 2;
  RELATIVE_DIRECTION_AROUND = 3;       // Fuzzy temporal window
}

// Semantic content search
message SemanticCue {
  string query = 1;                    // Natural language query
  float fuzzy_threshold = 2;           // Fuzzy matching threshold
  repeated string required_tags = 3;    // Must have these tags
  repeated string excluded_tags = 4;    // Must not have these tags
}

// Pattern completion cue
message PatternCue {
  repeated PartialMemory fragments = 1; // Partial memories to complete
  float completion_threshold = 2;       // Pattern completion threshold
}

// Partial memory for pattern completion
message PartialMemory {
  map<string, string> known_fields = 1;  // Known field values
  repeated string missing_fields = 2;    // Fields to reconstruct
  Confidence fragment_confidence = 3;    // Confidence in known fields
}

// ============================================================================
// Memory Operations Results - Rich response types
// ============================================================================

// Result of memory recall operation
message RecallResult {
  repeated Memory memories = 1;        // Retrieved memories
  Confidence recall_confidence = 2;    // Overall recall confidence
  RecallMetadata metadata = 3;         // Recall statistics
}

// Metadata about recall process
message RecallMetadata {
  int32 total_activated = 1;           // Total memories activated
  int32 above_threshold = 2;           // Memories above confidence threshold
  float avg_activation = 3;            // Average activation level
  int64 recall_time_ms = 4;            // Time taken for recall
  repeated string activation_path = 5;  // Spreading activation path
}

// Result of memory storage operation
message StoreResult {
  string memory_id = 1;                // Stored memory ID
  Confidence storage_confidence = 2;   // Storage success confidence
  bool consolidated = 3;               // Whether immediately consolidated
  repeated string linked_memories = 4;  // Auto-linked related memories
}

// ============================================================================
// Cognitive Errors - Educational error messages
// ============================================================================

// CognitiveError provides rich error context following cognitive ergonomics.
// Every error includes suggestion and example for tired/stressed developers.
message CognitiveError {
  string summary = 1;                  // One-line error summary
  string context = 2;                  // What was expected vs actual
  string suggestion = 3;               // How to fix it
  string example = 4;                  // Code example of correct usage
  Confidence error_confidence = 5;     // Confidence in error diagnosis
  repeated string similar_memories = 6; // "Did you mean?" suggestions
}

// Embedding wrapper for use in oneof fields
message Embedding {
  repeated float values = 1;
}