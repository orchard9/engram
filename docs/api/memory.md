# Memory Operations

## Store a Memory

Store new information in Engram's cognitive memory system.

**Endpoint:** `POST /api/v1/memories/remember`

### Request

```json
{
  "content": "The capital of France is Paris",
  "confidence": 0.9
}
```

### Response

```json
{
  "memory_id": "mem_a1b2c3d4",
  "storage_confidence": {
    "value": 0.9,
    "category": "High",
    "reasoning": "Strong factual content with clear semantic meaning"
  },
  "consolidation_state": "Recent",
  "auto_links": [],
  "system_message": "Memory successfully stored with high confidence"
}
```

## Recall Memories

Search and retrieve memories using natural language queries.

**Endpoint:** `GET /api/v1/memories/recall?query={search_terms}`

### Parameters

- `query` (required) - Search terms or question
- `confidence_threshold` (optional) - Minimum confidence (0.0-1.0)
- `max_results` (optional) - Maximum number of results (default: 10)

### Example

```bash
curl "http://localhost:7432/api/v1/memories/recall?query=capital+France"
```

### Response

```json
{
  "memories": {
    "vivid": [
      {
        "id": "mem_a1b2c3d4",
        "content": "The capital of France is Paris",
        "confidence": {
          "value": 0.9,
          "category": "High",
          "reasoning": "Direct factual match"
        },
        "activation_level": 0.95,
        "similarity_score": 0.92,
        "retrieval_path": "Direct similarity match",
        "last_access": "2025-09-17T01:30:00Z"
      }
    ],
    "associated": [],
    "reconstructed": []
  },
  "recall_confidence": {
    "value": 0.9,
    "category": "High",
    "reasoning": "Strong query-memory alignment"
  }
}
```

## Pattern Recognition

Recognize patterns in provided content.

**Endpoint:** `POST /api/v1/memories/recognize`

### Request

```json
{
  "content": "Paris is a beautiful city in Europe",
  "context": "geographical"
}
```

### Response

```json
{
  "recognized_patterns": [
    {
      "pattern_type": "Location",
      "confidence": 0.85,
      "description": "Geographic location reference"
    }
  ],
  "system_message": "Pattern recognition completed successfully"
}
```
