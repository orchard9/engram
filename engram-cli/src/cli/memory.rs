//! Memory operations for the CLI

use anyhow::Result;
use engram_core::MemorySpaceId;
use serde_json::{Value, json};
use std::time::Instant;
use tracing::{error, info};

/// Print formatted memory result
pub fn print_memory_result(memory: &Value) {
    if let Some(obj) = memory.as_object() {
        if let Some(id) = obj.get("id").and_then(|v| v.as_str()) {
            println!("üß† Memory ID: {id}");
        }

        if let Some(content) = obj.get("content").and_then(|v| v.as_str()) {
            println!("üìù Content: {content}");
        }

        if let Some(confidence) = obj.get("confidence").and_then(serde_json::Value::as_f64) {
            let confidence_bar = "‚ñà".repeat((confidence * 10.0).clamp(0.0, 10.0) as usize);
            println!(
                "üéØ Confidence: {:.1}% {}",
                confidence * 100.0,
                confidence_bar
            );
        }

        if let Some(timestamp) = obj.get("timestamp").and_then(|v| v.as_str()) {
            println!("‚è∞ Created: {timestamp}");
        }

        if let Some(associations) = obj.get("associations").and_then(|v| v.as_array())
            && !associations.is_empty()
        {
            println!("üîó Associated memories: {}", associations.len());
            for (i, assoc) in associations.iter().take(3).enumerate() {
                if let Some(assoc_obj) = assoc.as_object()
                    && let Some(content) = assoc_obj.get("content").and_then(|v| v.as_str())
                {
                    let preview = if content.len() > 50 {
                        format!("{}...", &content[..47])
                    } else {
                        content.to_string()
                    };
                    println!("  {}. {}", i + 1, preview);
                }
            }
            if associations.len() > 3 {
                println!("  ... and {} more", associations.len() - 3);
            }
        }

        println!();
    } else {
        println!(
            "üìÑ Raw result: {}",
            serde_json::to_string_pretty(memory).unwrap_or_else(|_| "Invalid JSON".to_string())
        );
    }
}

/// Create a memory through HTTP API
///
/// # Errors
///
/// Returns error if HTTP request fails or server returns an error
pub async fn create_memory(
    port: u16,
    content: String,
    confidence: Option<f64>,
    space_id: &MemorySpaceId,
) -> Result<()> {
    let client = reqwest::Client::new();
    let url = format!("http://127.0.0.1:{port}/api/v1/memories");

    let mut payload = json!({
        "content": content
    });

    if let Some(conf) = confidence {
        payload["confidence"] = json!(conf);
    }

    info!("üîÑ Creating memory in space '{}'...", space_id.as_str());
    let start_time = Instant::now();

    let response = client
        .post(&url)
        .header("X-Engram-Memory-Space", space_id.as_str())
        .json(&payload)
        .send()
        .await?;

    let elapsed = start_time.elapsed();

    if response.status().is_success() {
        let memory: Value = response.json().await?;
        println!("‚úÖ Memory created successfully in {elapsed:?}");
        print_memory_result(&memory);
    } else {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        error!("‚ùå Failed to create memory: {}", error_text);
        return Err(anyhow::anyhow!("Failed to create memory: {}", error_text));
    }

    Ok(())
}

/// Retrieve a memory by ID
///
/// # Errors
///
/// Returns error if HTTP request fails or memory not found
pub async fn get_memory(port: u16, id: String, space_id: &MemorySpaceId) -> Result<()> {
    let client = reqwest::Client::new();
    let url = format!("http://127.0.0.1:{port}/api/v1/memories/{id}");

    info!("üîç Retrieving memory from space '{}'...", space_id.as_str());
    let start_time = Instant::now();

    let response = client
        .get(&url)
        .header("X-Engram-Memory-Space", space_id.as_str())
        .send()
        .await?;
    let elapsed = start_time.elapsed();

    if response.status().is_success() {
        let memory: Value = response.json().await?;
        println!("‚úÖ Memory retrieved in {elapsed:?}");
        print_memory_result(&memory);
    } else if response.status() == 404 {
        println!("‚ùå Memory not found with ID: {id}");
    } else {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        error!("‚ùå Failed to retrieve memory: {}", error_text);
        return Err(anyhow::anyhow!("Failed to retrieve memory: {}", error_text));
    }

    Ok(())
}

/// Search for memories
///
/// # Errors
///
/// Returns error if HTTP request fails or search fails
pub async fn search_memories(
    port: u16,
    query: String,
    limit: Option<usize>,
    space_id: &MemorySpaceId,
) -> Result<()> {
    let client = reqwest::Client::new();
    let mut url = format!(
        "http://127.0.0.1:{}/api/v1/memories/search?query={}",
        port,
        urlencoding::encode(&query)
    );

    if let Some(lim) = limit {
        use std::fmt::Write;
        let _ = write!(&mut url, "&limit={lim}");
    }

    info!("üîç Searching memories in space '{}'...", space_id.as_str());
    let start_time = Instant::now();

    let response = client
        .get(&url)
        .header("X-Engram-Memory-Space", space_id.as_str())
        .send()
        .await?;
    let elapsed = start_time.elapsed();

    if response.status().is_success() {
        let results: Value = response.json().await?;

        if let Some(memories) = results.get("memories").and_then(|v| v.as_array()) {
            println!("‚úÖ Found {} memories in {:?}", memories.len(), elapsed);

            if memories.is_empty() {
                println!("üîç No memories found matching: '{query}'");
                println!("üí° Try a different search term or create some memories first");
            } else {
                for (i, memory) in memories.iter().enumerate() {
                    println!("üìã Result {} of {}:", i + 1, memories.len());
                    print_memory_result(memory);
                }
            }
        } else {
            println!("‚ö†Ô∏è  Unexpected response format");
        }
    } else {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        error!("‚ùå Failed to search memories: {}", error_text);
        return Err(anyhow::anyhow!("Failed to search memories: {}", error_text));
    }

    Ok(())
}

/// List all memories
///
/// # Errors
///
/// Returns error if HTTP request fails or listing fails
pub async fn list_memories(port: u16, limit: Option<usize>, offset: Option<usize>) -> Result<()> {
    let client = reqwest::Client::new();
    let mut url = format!("http://127.0.0.1:{port}/api/v1/memories");

    let mut params = Vec::new();
    if let Some(lim) = limit {
        params.push(format!("limit={lim}"));
    }
    if let Some(off) = offset {
        params.push(format!("offset={off}"));
    }

    if !params.is_empty() {
        url.push('?');
        url.push_str(&params.join("&"));
    }

    info!("üìã Listing memories...");
    let start_time = Instant::now();

    let response = client.get(&url).send().await?;
    let elapsed = start_time.elapsed();

    if response.status().is_success() {
        let results: Value = response.json().await?;

        if let Some(memories) = results.get("memories").and_then(|v| v.as_array()) {
            println!("‚úÖ Retrieved {} memories in {:?}", memories.len(), elapsed);

            if memories.is_empty() {
                println!("üì≠ No memories found");
                println!(
                    "üí° Create your first memory with: engram memory create \"Your thought here\""
                );
            } else {
                for (i, memory) in memories.iter().enumerate() {
                    println!("üìã Memory {} of {}:", i + 1, memories.len());
                    print_memory_result(memory);
                }
            }

            // Show pagination info if applicable
            if let Some(total) = results.get("total").and_then(serde_json::Value::as_u64) {
                let offset = offset.unwrap_or(0);
                let limit = limit.unwrap_or(10);
                if total > (offset + memories.len()) as u64 {
                    println!("üìÑ Showing {} of {} total memories", memories.len(), total);
                    println!(
                        "üí° Use --offset {} --limit {} to see more",
                        offset + limit,
                        limit
                    );
                }
            }
        }
    } else {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        error!("‚ùå Failed to list memories: {}", error_text);
        return Err(anyhow::anyhow!("Failed to list memories: {}", error_text));
    }

    Ok(())
}

/// Delete a memory by ID
///
/// # Errors
///
/// Returns error if HTTP request fails or deletion fails
pub async fn delete_memory(port: u16, id: String) -> Result<()> {
    let client = reqwest::Client::new();
    let url = format!("http://127.0.0.1:{port}/api/v1/memories/{id}");

    info!("üóëÔ∏è  Deleting memory...");

    let response = client.delete(&url).send().await?;

    if response.status().is_success() {
        println!("‚úÖ Memory deleted successfully");
    } else if response.status() == 404 {
        println!("‚ùå Memory not found with ID: {id}");
    } else {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        error!("‚ùå Failed to delete memory: {}", error_text);
        return Err(anyhow::anyhow!("Failed to delete memory: {}", error_text));
    }

    Ok(())
}
