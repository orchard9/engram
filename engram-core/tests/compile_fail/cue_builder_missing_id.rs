//! Test that Cue cannot be built without an ID
//! 
//! This test validates that the typestate pattern prevents construction
//! of Cue without the required ID field.

use engram_core::memory::CueBuilder;
use engram_core::Confidence;

fn main() {
    // This should fail: cannot build Cue without setting ID first
    let _cue = CueBuilder::new()
        .embedding_search([0.1f32; 768], Confidence::MEDIUM)
        .build(); //~ ERROR: method `build` not found for this value
    //~| NOTE: ID is required before building Cue
    //~| HELP: Start with: CueBuilder::new().id("unique_cue_id".to_string())
    //~| HELP: Cue ID enables tracking and caching of query results
}