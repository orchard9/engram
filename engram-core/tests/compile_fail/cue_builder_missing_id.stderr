error[E0599]: no method named `embedding_search` found for struct `CueBuilder<engram_core::memory::cue_builder_states::NoId>` in the current scope
  --> tests/compile_fail/cue_builder_missing_id.rs:12:10
   |
11 |       let _cue = CueBuilder::new()
   |  ________________-
12 | |         .embedding_search([0.1f32; 768], Confidence::MEDIUM)
   | |         -^^^^^^^^^^^^^^^^ method not found in `CueBuilder<engram_core::memory::cue_builder_states::NoId>`
   | |_________|
   |
   |
   = note: the method was found for
           - `CueBuilder<NoCueType>`
